server/chat-plugins/helptickets-auto.ts

```diff
@@ -0,0 +1,670 @@
+					actions.set(toID(userid), {
+						action: result.action,
+						user: toID(userid),
+						result: result.result,
+						reason: `Pokemon name detected to be breaking rules - '${result.name}'`,
+						roomid: link,
+					});
+				}
+			}
+		}
+		if (actions.size) return actions;
+	},
+	async battleharassment(ticket) {
+		const urls = getBattleLinks(ticket.text[0]);
+		const actions = new Map<string, CheckerResult>();
+		for (const url of urls) {
+			const log = await getBattleLog(url);
+			if (!log) continue;
+			const messages: Record<string, string[]> = {};
+			for (const message of log.log) {
+				const [username, text] = Utils.splitFirst(message.slice(3), '|').map(f => f.trim());
+				const id = toID(username);
+				if (!id) continue;
+				if (!messages[id]) messages[id] = [];
+				messages[id].push(text);
+			}
+			for (const [id, messageList] of Object.entries(messages)) {
+				const {averages, classified} = await getMessageAverages(messageList);
+				const punishment = determinePunishment('battleharassment', averages, []);
+				if (!punishment) continue;
+				const existingPunishment = actions.get(id);
+				if (!existingPunishment || supersedes(punishment, existingPunishment.action)) {
+					actions.set(id, {
+						action: punishment,
+						user: toID(id),
+						result: averages,
+						reason: `Not following rules in battles`,
+						proof: urls.join(', '),
+					});
+				}
+				for (const result of classified) {
+					const curPunishment = determinePunishment('battleharassment', result, [], true);
+					if (!curPunishment) continue;
+					const exists = actions.get(id);
+					if (!exists || supersedes(curPunishment, exists.action)) {
+						actions.set(id, {
+							action: curPunishment,
+							user: toID(id),
+							result: averages,
+							reason: `Not following rules in battles`,
+							proof: urls.join(', '),
+						});
+					}
+				}
+			}
+		}
+		// ensure reasons are clear
+		const creatorWasPunished = actions.get(ticket.userid);
+		if (creatorWasPunished) {
+			let displayReason = 'You were punished for your behavior.';
+			if (actions.size !== 1) { // more than 1 was punished
+				displayReason += ` ${actions.size - 1} other(s) were also punished.`;
+			}
+			creatorWasPunished.displayReason = displayReason;
+		}
+
+		if (actions.size) return actions;
+	},
+	async pmharassment(ticket) {
+		const actions = new Map<string, CheckerResult>();
+		const targetId = toID(ticket.text[0]);
+		const creator = ticket.userid;
+		if (!Config.getpmlog) return;
+		const pmLog = await Config.getpmlog(targetId, creator) as {
+			message: string, from: string, to: string, timestamp: string,
+		}[];
+		const messages: Record<string, string[]> = {};
+		const ids = new Set<ID>();
+		// sort messages by user who sent them, also filter out old ones
+		for (const {from, message, timestamp} of pmLog) {
+			// ignore pmlogs more than 24h old
+			if ((Date.now() - new Date(timestamp).getTime()) > PMLOG_IGNORE_TIME) continue;
```

Comment:
Why the `new Date(...).getTime()`? Don't you already have the UNIX timestamp?

Reply:
No, timestamp is a string of `YYYY-MM-DD HH:MM:SS`.